Dynamic Dashboard: Stats and chart views for store performance card

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #12465 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This PR adds stats and chart views for the new store performance card. I reuse `StoreStatsPeriodViewModel` to retain most of the existing logic, and copied the checks for the custom range redacted view from the legacy view controllers.

Changes:
- Updated `DashboardView` and its hosting controller to handle navigation to the analytic hub and presenting the custom range alert from the performance card.
- Updated the performance card to:
  - Display the selected date for custom time ranges. We don't have design for how to display this, so I'm replacing the date range with the selected date. I tried displaying the date below the range but it looks odd - please feel free to suggest a different option if you have an idea.
  - Add a view for stat values and redacted views.
  - Add chart view using the new chart from #12467.
  - Add a button to navigate to analytics hub.
  - Track interactions with the chart and time range picker.
- Updated `StoreStatsChart` to:
  - disable selecting chart when there is no revenue. 
  - support deselect points on the chart.
  - update no revenue view color.
 
Due to the large number of changes, unit tests will be added in a separate PR.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Enable the feature flag `dynamicDashboard` and build the app.
- Ensure that store performance card is enabled on the dashboard screen.
- Confirm that the card works properly:
  - Switch different time ranges and confirm that the stats and chart are updated correctly.
  - Tap on the chart, confirm that tapping only works when there is any revenue. Confirm that the selected date and stats values are displayed upon tapping. Tapping the same point again should deselect it and the values and charts should be back to the unselected state.
  - Switch between different stores, confirm that the visitor and conversion stats are redacted for JCP sites and when authenticated without WPCom.
  - Switch to a WPCom store and select a custom range. Confirm that the visitor and conversion stats are redacted for ranges longer than 1 day but not for 1 day. Select any point on the chart, confirm that the visitor stats are not redacted for ranges longer than 1 day.
  - Tap on the redacted view for custom range, confirm that the alert explaining the reason is displayed.
  - Tap on View all store analytics, confirm that the app navigates to analytics hub.
  - Confirm that tracking events are sent properly when interacting with the chart and time range picker.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

No revenue | With revenue
--- | ---
<img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/ea4e82fc-fa0d-49d5-9011-9788671a88f5" width=320 /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/5533851/2dbd29ec-0c3e-4065-91b8-6b8030122de8" width=320 />


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Observe stat texts

itsmeichigo committed:
Display stat texts

itsmeichigo committed:
Update chart selection

itsmeichigo committed:
Add chart view

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Highlight selected stats values

itsmeichigo committed:
Load and save last time range

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo committed:
Handle view all analytics

itsmeichigo committed:
Update no revenue view color

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Add additional texts for custom range

itsmeichigo committed:
Add redacted view

itsmeichigo committed:
Handle tap for custom range redacted view

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Restore tracking for chart

itsmeichigo committed:
Add default value for usageTracksEventEmitter in DashboardViewModel

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Update xAxisStrideCount for this year tab

itsmeichigo committed:
Fix highlight state for revenue stats

itsmeichigo committed:
Clean up logic for syncStats

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/196">18.3</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino commented on WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModel.swift:
```diff
@@ -32,43 +44,103 @@ final class StorePerformanceViewModel: ObservableObject {
+
+    func didSelectStatsInterval(at index: Int?) {
+        periodViewModel?.selectedIntervalIndex = index
+        chartValueSelectedEventsSubject.send(())
```
It seems that the double brackets are not necessary here

itsmeichigo replied:
Updated this in 7c6bd2f1141683039926e783e9ffd83ca1751c70.

pmusolino commented on WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StorePerformanceViewModel.swift:
```diff
@@ -133,6 +234,93 @@ private extension StorePerformanceViewModel {
+                    analytics.track(event: .DashboardCustomRange.interacted())
+                }
+                usageTracksEventEmitter.interacted()
+            }.store(in: &subscriptions)
```
It seems that this code works without following the number of seconds declared here, it's like it will be executed just sometime, and this seems that (probably) it's generating the following issue, dragging the finger across the chart.

https://github.com/woocommerce/woocommerce-ios/assets/495617/7e00d274-745a-4ae8-b512-ed57c8dad294

itsmeichigo replied:
Good catch! I updated the handling for chart selection in e8661b29c9de0e63fe82c5d4b0a7f5c105f58954.

pmusolino commented on WooCommerce/Classes/ViewRelated/Dashboard/StoreStats/StoreStatsChart.swift:
```diff
@@ -141,6 +155,10 @@ private extension StoreStatsChart {
     }
 
     enum Constants {
+        static var noRevenueLineColor = Color(
+            light: Color.withColorStudio(name: .gray, shade: .shade5),
+            dark: Color.withColorStudio(name: .gray, shade: .shade80)
```
minor nit: It seems we already declared these colors also in `WooPaymentsDepositStatusDisplayDetails`, I wonder if we can declare them just one time.

itsmeichigo replied:
These are separate view and the colors are the same by chance - I don't think it's safe to reuse the colors in this case. I'll keep them as is.

pmusolino reviewed (approved):
Looks good to me, good job @itsmeichigo!

I just found some minor issues that can be related also to other PRs regarding Dynamic Dashboard. But not blocking for the moment, they could be fixed into future PRs.

The more serious issue is the one about the finder dragging, down in the comments.

#### Wrong alignment of the icons
It seems also that the size of the 3 dots icon is super small and hard to tap.
<img width="300" alt="wrong alignment" src="https://github.com/woocommerce/woocommerce-ios/assets/495617/c561d0e0-af00-42d9-b320-72bf24aa345d">

#### In landscape, during the pull to refresh, for some reason the title of the store will be "splitted" from the navigation bar, while previously the name of store will be hidden in landscape mode.

![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-16 at 22 01 33](https://github.com/woocommerce/woocommerce-ios/assets/495617/22a494da-d395-472a-a520-08f485997d4f)

#### It should not be possible to open the menus, while the content is redacted.

| Dates range button            |  More button |
:-------------------------:|:-------------------------:
![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-16 at 21 41 07](https://github.com/woocommerce/woocommerce-ios/assets/495617/b5116ce5-9b00-42ea-b210-71e68e829cc3) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-04-16 at 21 41 43](https://github.com/woocommerce/woocommerce-ios/assets/495617/62f1e312-6352-41f3-82c3-a9d677f336bc)

itsmeichigo committed:
Merge branch 'feat/12403-store-performance-card' into feat/12465-store-peformance-chart

itsmeichigo committed:
Make buttons larger and easier to tap

itsmeichigo committed:
Render store title only if vertical size class is regular

itsmeichigo committed:
Remove redundant brackets

itsmeichigo committed:
Handle chart selection after debounce

itsmeichigo commented:
Thanks @pmusolino for the reviews! 🙇 

> It seems also that the size of the 3 dots icon is super small and hard to tap.

The calendar icon seems to have padding originally so I'm not sure how to adjust the alignment. I updated the ellipsis icon size in b398d365ec6381533aa684593a1c9af92db49921.

> In landscape, during the pull to refresh, for some reason the title of the store will be "splitted" from the navigation bar, while previously the name of store will be hidden in landscape mode.

I fixed this by adding store title only when the vertical size class is regular in 1b1761f77543b101ae2d639380382615caf92aba.

> It should not be possible to open the menus, while the content is redacted.

Also fixed as part of b398d365ec6381533aa684593a1c9af92db49921.
